File: task-db3fcf/environment/app/lexer.py
"""Lexer for the MiniCalc language."""


class TokenType:
    INT = 'INT'
    IDENT = 'IDENT'
    PLUS = '+'
    MINUS = '-'
    STAR = '*'
    SLASH = '/'
    PERCENT = '%'
    EQ_EQ = '=='
    BANG_EQ = '!='
    LT = '<'
    LE = '<='
    GT = '>'
    GE = '>='
    AND = '&&'
    OR = '||'
    BANG = '!'
    ASSIGN = '='
    LPAREN = '('
    RPAREN = ')'
    LBRACE = '{'
    RBRACE = '}'
    SEMI = ';'
    COMMA = ','
    IF = 'if'
    ELSE = 'else'
    WHILE = 'while'
    FUNC = 'func'
    RETURN = 'return'
    PRINT = 'print'
    EOF = 'EOF'


class Token:
    def __init__(self, type_, value, line):
        self.type = type_
        self.value = value
        self.line = line

    def __repr__(self):
        return f'Token({self.type}, {self.value!r}, line={self.line})'


KEYWORDS = {'if', 'else', 'while', 'func', 'return', 'print'}


class LexerError(Exception):
    pass


class Lexer:
    def __init__(self, source):
        self.source = source
        self.pos = 0
        self.line = 1

    def _skip_ws_comments(self):
        while self.pos < len(self.source):
            c = self.source[self.pos]
            if c in ' \t\r':
                self.pos += 1
            elif c == '\n':
                self.pos += 1
                self.line += 1
            elif (self.pos + 1 < len(self.source) and
                  self.source[self.pos:self.pos + 2] == '//'):
                while self.pos < len(self.source) and self.source[self.pos] != '\n':
                    self.pos += 1
            else:
                break

    def next_token(self):
        self._skip_ws_comments()
        if self.pos >= len(self.source):
            return Token(TokenType.EOF, None, self.line)

        c = self.source[self.pos]

        # Two-character tokens (must check before single-char)
        if self.pos + 1 < len(self.source):
            two = self.source[self.pos:self.pos + 2]
            if two == '==':
                self.pos += 2
                return Token(TokenType.EQ_EQ, '==', self.line)
            if two == '!=':
                self.pos += 2
                return Token(TokenType.BANG_EQ, '!=', self.line)
            if two == '<=':
                self.pos += 2
                return Token(TokenType.LE, '<=', self.line)
            if two == '>=':
                self.pos += 2
                return Token(TokenType.GE, '>=', self.line)
            if two == '&&':
                self.pos += 2
                return Token(TokenType.AND, '&&', self.line)
            if two == '||':
                self.pos += 2
                return Token(TokenType.OR, '||', self.line)

        # Single-character tokens
        singles = {
            '+': TokenType.PLUS, '-': TokenType.MINUS, '*': TokenType.STAR,
            '/': TokenType.SLASH, '%': TokenType.PERCENT, '<': TokenType.LT,
            '>': TokenType.GT, '!': TokenType.BANG, '=': TokenType.ASSIGN,
            '(': TokenType.LPAREN, ')': TokenType.RPAREN,
            '{': TokenType.LBRACE, '}': TokenType.RBRACE,
            ';': TokenType.SEMI, ',': TokenType.COMMA,
        }
        if c in singles:
            self.pos += 1
            return Token(singles[c], c, self.line)

        # Integer literal
        if c.isdigit():
            start = self.pos
            while self.pos < len(self.source) and self.source[self.pos].isdigit():
                self.pos += 1
            return Token(TokenType.INT, int(self.source[start:self.pos]), self.line)

        # Identifier or keyword
        if c.isalpha() or c == '_':
            start = self.pos
            while (self.pos < len(self.source) and
                   (self.source[self.pos].isalnum() or self.source[self.pos] == '_')):
                self.pos += 1
            word = self.source[start:self.pos]
            if word in KEYWORDS:
                return Token(word, word, self.line)
            return Token(TokenType.IDENT, word, self.line)

        raise LexerError(f"Unexpected character '{c}' at line {self.line}")

    def tokenize(self):
        tokens = []
        while True:
            tok = self.next_token()
            tokens.append(tok)
            if tok.type == TokenType.EOF:
                break
        return tokens

File: bids-expression-evaluator/environment/app/src/lexer.ts

import { Token, TokenType } from './types';

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;

  while (pos < input.length) {
    // Skip whitespace
    if (/\s/.test(input[pos])) {
      pos++;
      continue;
    }

    // Numbers (integer or floating point)
    if (/\d/.test(input[pos]) || (input[pos] === '.' && pos + 1 < input.length && /\d/.test(input[pos + 1]))) {
      let num = '';
      while (pos < input.length && /[\d.]/.test(input[pos])) {
        num += input[pos++];
      }
      tokens.push({ type: TokenType.Number, value: num, position: pos });
      continue;
    }

    // String literals (single or double quoted)
    if (input[pos] === '"' || input[pos] === "'") {
      const quote = input[pos++];
      let str = '';
      while (pos < input.length && input[pos] !== quote) {
        if (input[pos] === '\\') {
          pos++;
          str += input[pos++];
        } else {
          str += input[pos++];
        }
      }
      pos++; // closing quote
      tokens.push({ type: TokenType.String, value: str, position: pos });
      continue;
    }

    // Identifiers and keywords
    if (/[a-zA-Z_]/.test(input[pos])) {
      let ident = '';
      while (pos < input.length && /[a-zA-Z0-9_]/.test(input[pos])) {
        ident += input[pos++];
      }
      if (ident === 'true' || ident === 'false') {
        tokens.push({ type: TokenType.Boolean, value: ident, position: pos });
      } else if (ident === 'null') {
        tokens.push({ type: TokenType.Null, value: ident, position: pos });
      } else if (ident === 'in') {
        tokens.push({ type: TokenType.In, value: ident, position: pos });
      } else {
        tokens.push({ type: TokenType.Identifier, value: ident, position: pos });
      }
      continue;
    }

    // Two-character operators
    if (pos + 1 < input.length) {
      const two = input[pos] + input[pos + 1];
      if (two === '==') {
        tokens.push({ type: TokenType.EqualEqual, value: '==', position: pos });
        pos += 2;
        continue;
      }
      if (two === '<=') {
        tokens.push({ type: TokenType.LessEqual, value: '<=', position: pos });
        pos += 2;
        continue;
      }
      if (two === '>=') {
        tokens.push({ type: TokenType.GreaterEqual, value: '>=', position: pos });
        pos += 2;
        continue;
      }
      if (two === '&&') {
        tokens.push({ type: TokenType.And, value: '&&', position: pos });
        pos += 2;
        continue;
      }
      if (two === '||') {
        tokens.push({ type: TokenType.Or, value: '||', position: pos });
        pos += 2;
        continue;
      }
    }

    // Single-character tokens
    switch (input[pos]) {
      case '(':
        tokens.push({ type: TokenType.LeftParen, value: '(', position: pos });
        break;
      case ')':
        tokens.push({ type: TokenType.RightParen, value: ')', position: pos });
        break;
      case '[':
        tokens.push({ type: TokenType.LeftBracket, value: '[', position: pos });
        break;
      case ']':
        tokens.push({ type: TokenType.RightBracket, value: ']', position: pos });
        break;
      case '{':
        tokens.push({ type: TokenType.LeftBrace, value: '{', position: pos });
        break;
      case '}':
        tokens.push({ type: TokenType.RightBrace, value: '}', position: pos });
        break;
      case ',':
        tokens.push({ type: TokenType.Comma, value: ',', position: pos });
        break;
      case '.':
        tokens.push({ type: TokenType.Dot, value: '.', position: pos });
        break;
      case '+':
        tokens.push({ type: TokenType.Plus, value: '+', position: pos });
        break;
      case '-':
        tokens.push({ type: TokenType.Minus, value: '-', position: pos });
        break;
      case '*':
        tokens.push({ type: TokenType.Star, value: '*', position: pos });
        break;
      case '/':
        tokens.push({ type: TokenType.Slash, value: '/', position: pos });
        break;
      case '%':
        tokens.push({ type: TokenType.Percent, value: '%', position: pos });
        break;
      case '!':
        tokens.push({ type: TokenType.Bang, value: '!', position: pos });
        break;
      case '=':
        tokens.push({ type: TokenType.Equal, value: '=', position: pos });
        break;
      case '<':
        tokens.push({ type: TokenType.Less, value: '<', position: pos });
        break;
      case '>':
        tokens.push({ type: TokenType.Greater, value: '>', position: pos });
        break;
      default:
        throw new Error(`Unexpected character: ${input[pos]} at position ${pos}`);
    }
    pos++;
  }

  tokens.push({ type: TokenType.EOF, value: '', position: pos });
  return tokens;
}
